libraries/AP_Module/AP_Module.cpp

```diff
@@ -0,0 +1,194 @@
+/// -*- tab-width: 4; Mode: C++; c-basic-offset: 4; indent-tabs-mode: nil -*-
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+  support for external modules
+ */
+
+#include <stdio.h>
+#include <dirent.h>
+#if defined(HAVE_LIBDL)
+#include <dlfcn.h>
+#endif
+#include <AP_Module/AP_Module.h>
+#include <AP_Module/AP_Module_Structures.h>
+
+struct AP_Module::hook_list *AP_Module::hooks[NUM_HOOKS];
+
+const char *AP_Module::hook_names[AP_Module::NUM_HOOKS] = {
+    "hook_setup_start",
+    "hook_setup_complete",
+    "hook_AHRS_update",
```

Comment:
I've changed it to ap_hook. I don't think it should be ap_module as its not really part of the AP_Module class.